### ISSUE TITLE
Zelda Majora's Mask: 2 Suggested options to have in GLideN64.custom.ini file

### ISSUE THREAD

**Issue by IronYoshi64** [gonetz/GLideN64#2315]
Hey! I been messing around with Zelda Majora's Mask a lot, and I been noticing for a while 2 small things that I really believe should be on by default for the plugin.

These are the options I recommend to be on for the game:

**[ZELDA%20MAJORA'S%20MASK]
frameBufferEmulation\copyAuxToRDRAM=1
generalEmulation\enableLOD=0**

I have played with these for a while and I did not notice any slowdowns, regressions, or glitches with these On so I think they are safe to enable.

generalEmulation\enableLOD=0 ----> This fixes the Southern Swamp Restored state with the bugged pinkish-purple tiles over the grass textures. Here are 2 pics to show what I mean:

Swamp restored - incorrect grass:
![GLideN64_ZELDA_MAJORA'S_MASK_000](https://user-images.githubusercontent.com/57850880/89311736-fd576880-d63b-11ea-97bf-02632f4204b1.png)

Swamp restored - correct grass:
![GLideN64_ZELDA_MAJORA'S_MASK_001](https://user-images.githubusercontent.com/57850880/89311781-0f390b80-d63c-11ea-8f7b-49b4043bb379.png)

frameBufferEmulation\copyAuxToRDRAM=1 ----> This fixes the quick fade in and fade out effect when opening and closing the Bombers Notebook. Here is a pic of the glitchy quick transition screen:

Bombers notebook transition effect messed up, should be simple black screen fade:
![GLideN64_ZELDA_MAJORA'S_MASK_002](https://user-images.githubusercontent.com/57850880/89312155-85d60900-d63c-11ea-8f64-3bc4f2cdb03a.png)

**Comment by standard-two-simplex** [comment]
@IronYoshi64 They are two separated issues so it would be better to create two issue threads for separated discussion.

About the pink grass issue, disabling LOD emulation is definitely not a solution, as the N64 uses LOD for different effects. It may hide this issue but that is all.

Do you know if it has been there forever? I can't recall seeing the issue before, but I'm getting it with Public Release 4 too. Can you confirm it is this old?

**Comment by IronYoshi64** [comment]
Like I said i spent very long amounts of time playing through and looking around for things that are weird, and for me, turning off the LOD thing doesn't affect the game is any way what so ever that I can notice, besides fix that swamp grass thing.

Also yeah, this has been an issue ever since Public Release 4.0. I downloaded PR 1.2, 2.0 and 3.0, went back to the area, and sure enough, it is completely not an issue in those Older builds, so something in 4.0 broke it...  also used only the defaults for each PR build.

**Comment by standard-two-simplex** [comment]
I hacked the lod code to display each lod of the texture (max level is 3). This are the pictures I got.

<table>
  <tr>
    <td>uTex0 - lod0 (lod_tile=0 ??) </td>
    <td>uTex1 - lod0 (lod_tile=1 ??)</td>
  </tr>
 <tr>
    <td><img src="https://user-images.githubusercontent.com/12202580/89511314-07ab6b00-d7d2-11ea-9f64-aba6b52f842f.png" width="320px" /> </td>
    <td><img src="https://user-images.githubusercontent.com/12202580/89511318-0a0dc500-d7d2-11ea-99a4-d133edf12493.png" width="320px" /></td>
  </tr>
  <tr>
    <td>uTex1 - lod1 (lod_tile=2 ??) </td>
    <td>uTex1 - lod2 (lod_tile=3 ??)</td>
  </tr>
 <tr>
    <td><img src="https://user-images.githubusercontent.com/12202580/89511326-0bd78880-d7d2-11ea-990e-8e60b7259bec.png" width="320px" /> </td>
    <td><img src="https://user-images.githubusercontent.com/12202580/89511329-0c701f00-d7d2-11ea-915e-f6022db137f0.png" width="320px" /></td>
  </tr>
</table>

The two most detailed maps contain a normal grass texture and display the picture correctly. The other two contain some strange texture, so the distant objects show garbage. In theory, the maximum level of detail is 3 so those "garbage" textures are supposed to be used.

One curious thing is that if I dump textures and load them without edition, I get the following picture.
<img src="https://user-images.githubusercontent.com/12202580/89515011-de410e00-d7d6-11ea-8693-643f5c2e41f8.png" width="240px"/>

This is the texture causing it (hash ZELDA MAJORA'S MASK#29905D74#0#2_all). This is amplified, click on it to see the original size.

<img src="https://user-images.githubusercontent.com/12202580/89514937-cbc6d480-d7d6-11ea-98eb-cb8acce2781c.png" width="64px">

@gonetz Is it possible to dump all mip-maps of the texture? Why does it dump/load that texture as most detailed when `uTex0` seems to contain another one?

**Comment by standard-two-simplex** [comment]
This is what a N64 looks like. It seems the correct behaviour is that the distant textures fade away.

https://youtu.be/PaZaZMry1_A?t=3350

**Comment by gonetz** [comment]
> About the pink grass issue, disabling LOD emulation is definitely not a solution, as the N64 uses LOD for different effects. It may hide this issue but that is all.

Yes, the right way to solve the issue is fixing LOD code.


> Is it possible to dump all mip-maps of the texture?

As I remember, no. Actually it is not necessary, since usually mip-maps are scaled down versions of the most detailed tile.


> Why does it dump/load that texture as most detailed when `uTex0` seems to contain another one?

I need to investigate, but it is probably the "detailed" texture case. N64 LOD mip-mapping has interesting feature: detailed texture. This is special texture, which usually has a format different from the format of the main texture. For example, main tiles have format CI, but detailed texture has I format. This texture is used when magnifying, that is when polygon is larger than texture. In that case "detailed" texture is blended with largest tile in mip-mapped tile set to add more details to the image.
When that "detailed" texture is used (detailed mode is enabled and  magnifying), it is loaded into T0 and most detailed tile of the main texture is loaded into T1.

**Comment by standard-two-simplex** [comment]
> usually mip-maps are scaled down versions of the most detailed tile.

I wanted to make sure, because I suspect this might be one of those unusual cases.

> it is probably the "detailed" texture case

I thought the same, but the detail flag was unset when I checked. Check just in case, to make sure that I didn't make a mistake.

I started refactoring LOD code because I realised one thing was off. Perhaps this issue is related. Funnily, I ended up investigating the Donkey Kong issue because I thought it was a LOD issue. I will create a separate issue thread explaining the details.
